File: 033.Search_in_Rotated_Sorted_Array/Solution.py
class Solution(object):
    def search(self, nums, target):
        low = 0
        high = len(nums)-1
        while low <= high:
            mid = (low + high) // 2
            print(low,high,mid)

            if nums[mid] == target:
                return mid
            elif high==low:
                return -1
            elif nums[mid] <= nums[low] and nums[mid] <= nums[high] and nums[mid-1] >= nums[mid]:#mid is pivot

                if target <= nums[high]:
                    low = mid + 1
                else:
                    high = mid - 1
            elif nums[mid] > nums[mid-1] and nums[high] > nums[mid]: #pivot to left of mid\
                if nums[mid] > nums[low]: #pivot at start index

                    if target < nums[mid]:
                        high = mid - 1
                    else:
                        low = mid + 1
                else:
                    if target > nums[mid] and target <= nums[high]:
                        low = mid + 1
                    elif target < nums[mid] or target >= nums[low]:
                        high = mid - 1
                    else:
                        return -1
            elif nums[mid] >= nums[low] and nums[high] <= nums[mid]: #pivot to right of mid
                if target <= nums[high] or target > nums[mid] :
                    low = mid + 1
                else:
                    high = mid - 1
            else:
                    return -1
        return -1

File: 033.Search_in_Rotated_Sorted_Array/Solution.java
class Solution {
    public int search(int[] nums, int target) {
    int lo = 0, hi = nums.length - 1;
    while (lo <= hi) {
        int mid = lo + (hi - lo) / 2;
        int num = nums[mid]; 
        //nums [ mid ] 和 target 在同一段
        if ((nums[mid] < nums[0]) == (target < nums[0])) {
            num = nums[mid];
        //nums [ mid ] 和 target 不在同一段，同时还要考虑下变成 -inf 还是 inf。
        } else {
            num = target < nums[0] ? Integer.MIN_VALUE : Integer.MAX_VALUE;
        }

        if (num < target)
            lo = mid + 1;
        else if (num > target)
            hi = mid - 1;
        else
            return mid;
    }
    return -1;
}
}